llvm/lib/IR/Instruction.cpp

```diff
@@ -554,6 +554,20 @@ void Instruction::dropUBImplyingAttrsAndMetadata() {
   dropUBImplyingAttrsAndUnknownMetadata(KnownIDs);
 }
 
+bool Instruction::hasUBImplyingAttrs() const {
+  auto *CB = dyn_cast<CallBase>(this);
+  if (!CB)
+    return false;
+  // For call instructions, we also need to check parameter and return
+  // attributes that are can cause UB.
+  for (unsigned ArgNo = 0; ArgNo < CB->arg_size(); ArgNo++)
+    if (CB->isPassingUndefUB(ArgNo))
+      return true;
+  return CB->hasRetAttr(Attribute::NoUndef) ||
+         CB->getRetDereferenceableBytes() > 0 ||
+         CB->getRetDereferenceableOrNullBytes() > 0;
```

Comment:
Using hasRetAttr(Attribute::Dereferenceable) is more efficient (avoids looking up the actual attribute).

Though it would be nicer to base these checks on getUBImplyingAttributes() instead.